# PREVIO

## Importar librerias

In [86]:
import pandas as pd
import numpy as np
import re

## Primera toma de contactos con el DataSet

In [45]:
sharkattacks = pd.read_csv("../project-pandas/GSAF5.csv", engine = "python")

In [46]:
sharkattacks.head()

Case Number       Date  Year        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   
4    2016.09.15  16-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex   ...  \
0  New Smyrna Beach, Volusia County  Surfing             male    M  ...   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano    M  ...   
2  New Smyrna Beach, Volusia County  Surfing             male    M  ...   
3                  Thirteenth Beach  Surfing  Rory Angiolella    M  ...   
4                       Bells Beach  Surfing             male    M  ...   

    Species        Investigator or Source                        pdf  \
0        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1        NaN  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3        NaN           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4  2 m shark           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.c   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.b   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.a   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.17   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.16   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0  2016.09.18.c           5993         NaN         NaN  
1  2016.09.18.b           5992         NaN         NaN  
2  2016.09.18.a           5991         NaN         NaN  
3    2016.09.17           5990         NaN         NaN  
4    2016.09.15           5989         NaN         NaN  

[5 rows x 24 columns]

In [47]:
sharkattacks.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [48]:
sharkattacks.shape

(5992, 24)

Tenemos un dataset amplio con 5992 filas y 24 columnas. En la primera toma de contacto con el DataSet vemos que hay algunos parámetros del DataSet que no parecen tener mucha relevancia por la cantidad de valores repetidos y valores NaN que observo.

Posibles procedimientos a aplicar:

- Observamos en Case Number tenemos unos índices al final que podremos limpiar.
- También observo como case number y date pese a tener formatos diferentes, dan información similar.
- Las columnas unnamed a primera vista no tienen valores relevantes.
- Columnas como pdf, href, investigator or source no aportan información relevante a priori

In [49]:
sharkattacks.isnull().sum()

Case Number                  0
Date                         0
Year                         0
Type                         0
Country                     43
Area                       402
Location                   496
Activity                   527
Name                       200
Sex                        567
Age                       2681
Injury                      27
Fatal (Y/N)                 19
Time                      3213
Species                   2934
Investigator or Source      15
pdf                          0
href formula                 1
href                         3
Case Number.1                0
Case Number.2                0
original order               0
Unnamed: 22               5991
Unnamed: 23               5990
dtype: int64

Podemos ver como varias de las columnas que tiene el DataSet tienen aproximadamente el 50% de sus celdas con valores nulos.

Véase Time, Age, Species o las columnas Unnamed

# MANIPULACIÓN Y LIMPIEZA

Podemos crear una función con el propósito de limpiar los campos en los que nos encontremos puntos, comas, parentesis o elementos que no nos aportan información relevante 

In [84]:
def clean_col(items):
column_cleaned = []
    for item in items:
        item = item.replace("(", "" )
        item = item.replace(")", "" )
        item = item.replace("/", "" )
        item = item.replace(".", ' ')
        item = item.replace(" ", '_')
        
    
        column_cleaned.append(item)
    return column_cleaned

IndentationError: expected an indented block (<ipython-input-84-8b15e7997bd4>, line 2)

Visto que las columnas Unnamed no tienen valores relevantes en prácticamente ninguna celda. Podemos proceder a eliminarlas del DataSet

In [59]:
sharkattacks.drop(['Unnamed: 22','Unnamed: 23'], axis=1)

Case Number         Date  Year        Type             Country  \
0     2016.09.18.c    18-Sep-16  2016  Unprovoked                 USA   
1     2016.09.18.b    18-Sep-16  2016  Unprovoked                 USA   
2     2016.09.18.a    18-Sep-16  2016  Unprovoked                 USA   
3       2016.09.17    17-Sep-16  2016  Unprovoked           AUSTRALIA   
4       2016.09.15    16-Sep-16  2016  Unprovoked           AUSTRALIA   
...            ...          ...   ...         ...                 ...   
5987       ND.0005  Before 1903     0  Unprovoked           AUSTRALIA   
5988       ND.0004  Before 1903     0  Unprovoked           AUSTRALIA   
5989       ND.0003    1900-1905     0  Unprovoked                 USA   
5990       ND.0002    1883-1889     0  Unprovoked              PANAMA   
5991       ND.0001    1845-1853     0  Unprovoked  CEYLON (SRI LANKA)   

                   Area                             Location      Activity  \
0               Florida     New Smyrna Beach, Volusia County       Surfing   
1               Florida     New Smyrna Beach, Volusia County       Surfing   
2               Florida     New Smyrna Beach, Volusia County       Surfing   
3              Victoria                     Thirteenth Beach       Surfing   
4              Victoria                          Bells Beach       Surfing   
...                 ...                                  ...           ...   
5987  Western Australia                          Roebuck Bay        Diving   
5988  Western Australia                                  NaN  Pearl diving   
5989     North Carolina                       Ocracoke Inlet      Swimming   
5990                NaN                 Panama Bay 8�N, 79�W           NaN   
5991   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       Name Sex   ... Fatal (Y/N)   Time   Species   \
0                      male    M  ...           N  13h00        NaN   
1            Chucky Luciano    M  ...           N  11h00        NaN   
2                      male    M  ...           N  10h43        NaN   
3           Rory Angiolella    M  ...           N    NaN        NaN   
4                      male    M  ...           N    NaN  2 m shark   
...                     ...  ...  ...         ...    ...        ...   
5987                   male    M  ...           Y    NaN        NaN   
5988                  Ahmun    M  ...           Y    NaN        NaN   
5989  Coast Guard personnel    M  ...           Y    NaN        NaN   
5990        Jules Patterson    M  ...           Y    NaN        NaN   
5991                   male    M  ...           Y    NaN        NaN   

                     Investigator or Source                             pdf  \
0               Orlando Sentinel, 9/19/2016            2016.09.18.c-NSB.pdf   
1               Orlando Sentinel, 9/19/2016        2016.09.18.b-Luciano.pdf   
2               Orlando Sentinel, 9/19/2016            2016.09.18.a-NSB.pdf   
3                        The Age, 9/18/2016       2016.09.17-Angiolella.pdf   
4                        The Age, 9/16/2016       2016.09.16-BellsBeach.pdf   
...                                     ...                             ...   
5987       H. Taunton; N. Bartlett,  p. 234          ND-0005-RoebuckBay.pdf   
5988  H. Taunton; N. Bartlett,  pp. 233-234               ND-0004-Ahmun.pdf   
5989   F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   
5990                    The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
5991                             S.W. Baker              ND-0001-Ceylon.pdf   

                                           href formula  \
0     http://sharkattackfile.net/spreadsheets/pdf_di...   
1     http://sharkattackfile.net/spreadsheets/pdf_di...   
2     http://sharkattackfile.net/spreadsheets/pdf_di...   
3     http://sharkattackfile.net/spreadsheets/pdf_di...   
4     http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                  

Podemos definir una funcion para todos aquellos NaN que nos encontremos en algunas columnas donde he encontrado más, pasar otro parámetro diferente.

In [64]:
def change_NaN (column):
    
    unk = column.fillna ('Unknown')
    
    return unk

sharkattacks['Time'] = change_NaN (sharkattacks['Time'])
sharkattacks['Age'] = change_NaN (sharkattacks['Age'])
sharkattacks['Area'] = change_NaN (sharkattacks['Area'])
sharkattacks['Location'] = change_NaN (sharkattacks['Location'])
sharkattacks['Activity'] = change_NaN (sharkattacks['Activity'])



Podemos pasar a normalizar los nombres de los países ya que tenemos filas con elementos que sobran como (_) o (/) o puntos (.) y limpiar los campos

In [87]:
pd.unique(sharkattacks['Country'])

array(['USA', 'AUSTRALIA', 'NEW CALEDONIA', 'REUNION', 'BAHAMAS', 'SPAIN',
       'CHINA', 'JAPAN', 'COLUMBIA', 'SOUTH AFRICA', 'EGYPT',
       'NEW ZEALAND', 'INDONESIA', 'FRENCH POLYNESIA', 'CAPE VERDE',
       'Fiji', 'BRAZIL', 'DOMINICAN REPUBLIC', 'CAYMAN ISLANDS',
       'UNITED ARAB EMIRATES', 'ARUBA', 'MOZAMBIQUE', 'THAILAND', 'FIJI',
       'PUERTO RICO', 'ITALY', 'MEXICO', 'ATLANTIC OCEAN', 'GREECE',
       'MAURITIUS', nan, 'ST. MARTIN', 'FRANCE', 'ECUADOR',
       'PAPUA NEW GUINEA', 'TRINIDAD & TOBAGO', 'KIRIBATI', 'ISRAEL',
       'DIEGO GARCIA', 'TAIWAN', 'JAMAICA', 'PALESTINIAN TERRITORIES',
       'GUAM', 'SEYCHELLES', 'BELIZE', 'PHILIPPINES', 'NIGERIA', 'TONGA',
       'SCOTLAND', 'CANADA', 'CROATIA', 'SAUDI ARABIA', 'CHILE',
       'ANTIGUA', 'KENYA', 'RUSSIA', 'TURKS & CAICOS', 'COSTA RICA',
       'UNITED KINGDOM', 'MALAYSIA', 'UNITED ARAB EMIRATES (UAE)',
       'SAMOA', 'AZORES', 'SOLOMON ISLANDS', 'SOUTH KOREA', 'MALTA',
       'VIETNAM', 'MADAGASCAR', 'PANAMA',

In [88]:
sharkattacks['Country'].value_counts()


USA                               2116
AUSTRALIA                         1279
SOUTH AFRICA                       565
PAPUA NEW GUINEA                   133
NEW ZEALAND                        125
                                  ... 
SOLOMON ISLANDS / VANUATU            1
CYPRUS                               1
MONACO                               1
FEDERATED STATES OF MICRONESIA       1
Coast of AFRICA                      1
Name: Country, Length: 203, dtype: int64

In [93]:
sharkattacks['Country'] = [(str(i).title()).replace("_"," ") 
                           for i in clean_col(sharkattacks['Country'])]
                            

AttributeError: 'float' object has no attribute 'replace'

Visto el dataset con detenimiento, podemos descartar más columnas que no nos hacen falta y podemos simplificar el dataset obteniendo resultados similares

In [101]:
sharkattacks = sharkattacks.drop(['Date','Name', 'Injury', 'Fatal (Y/N)', 'Time',
                                    'Investigator Or Source', 'pdf', 'href formula', 'href',
                                    'Case Number.1', 'Case Number.2', 'original order'], axis=1)

KeyError: "['Date' 'Name' 'Injury' 'Fatal (Y/N)' 'Time' 'Investigator Or Source'\n 'pdf' 'href formula' 'href' 'Case Number.1' 'Case Number.2'\n 'original order'] not found in axis"

Podemos crear una columna nueva asociando los años a decadas (1910->10s,1920->20s,1930->30s...)
Haria falta aplicarlo a todos los años, solo aplicare una funcion para los años comprendidos entre 1900 y 2016)

In [109]:
pd.unique(sharkattacks['Year'])

array([2016, 2015, 2014, 2013, 2012, 2011, 2010, 2009, 2008, 2007, 2006,
       2005, 2004, 2003, 2002, 2001, 2000, 1999, 1998, 1997, 1996, 1995,
       1984, 1994, 1993, 1992, 1991, 1990, 1989, 1969, 1988, 1987, 1986,
       1985, 1983, 1982, 1981, 1980, 1979, 1978, 1977, 1976, 1975, 1974,
       1973, 1972, 1971, 1970, 1968, 1967, 1966, 1965, 1964, 1963, 1962,
       1961, 1960, 1959, 1958, 1957, 1956, 1955, 1954, 1953, 1952, 1951,
       1950, 1949, 1948, 1848, 1947, 1946, 1945, 1944, 1943, 1942, 1941,
       1940, 1939, 1938, 1937, 1936, 1935, 1934, 1933, 1932, 1931, 1930,
       1929, 1928, 1927, 1926, 1925, 1924, 1923, 1922, 1921, 1920, 1919,
       1918, 1917, 1916, 1915, 1914, 1913, 1912, 1911, 1910, 1909, 1908,
       1907, 1906, 1905, 1904, 1903, 1902, 1901, 1900, 1899, 1898, 1897,
       1896, 1895, 1894, 1893, 1892, 1891, 1890, 1889, 1888, 1887, 1886,
       1885, 1884, 1883, 1882, 1881, 1880, 1879, 1878, 1877, 1876, 1875,
       1874, 1873, 1872, 1871, 1870, 1869, 1868, 18

In [110]:
decades = []
for year in sharkattacks['Year']:
    
    if year <= 1910 and year > 1900:
        decades.append('910s')
    if year <= 1920 and year > 1910:
        decades.append('920s')
    if year <= 1930 and year > 1920:
        decades.append('930s')
    if year <= 1940 and year > 1930:
        decades.append('940s')
    if year <= 1950 and year > 1940:
        decades.append('950s')
    if year <= 1960 and year > 1950:
        decades.append('960s')
    if year <= 1970 and year > 1960:
        decades.append('970s')
    if year <= 1980 and year > 1970:
        decades.append('980s')
    if year <= 1990 and year > 1980:
        decades.append('990s')
    if year <= 2000 and year > 1990:
        decades.append('2000s')
    if year <= 2010 and year > 2000:
        decades.append('2010s')
    
sharkattacks['Decades'] = Decades

NameError: name 'Decades' is not defined

Podemos pasar a reordenar las columnas para tener mejor lectura de los datos

In [96]:
newcolumns = ['Country','Area','Location','Activity','Type','Case Number','Year','Age']
sharkattacks = sharkattacks[newcolumns]

In [98]:
sharkattacks

Country               Area  \
0                    USA            Florida   
1                    USA            Florida   
2                    USA            Florida   
3              AUSTRALIA           Victoria   
4              AUSTRALIA           Victoria   
...                  ...                ...   
5987           AUSTRALIA  Western Australia   
5988           AUSTRALIA  Western Australia   
5989                 USA     North Carolina   
5990              PANAMA            Unknown   
5991  CEYLON (SRI LANKA)   Eastern Province   

                                 Location      Activity        Type  \
0        New Smyrna Beach, Volusia County       Surfing  Unprovoked   
1        New Smyrna Beach, Volusia County       Surfing  Unprovoked   
2        New Smyrna Beach, Volusia County       Surfing  Unprovoked   
3                        Thirteenth Beach       Surfing  Unprovoked   
4                             Bells Beach       Surfing  Unprovoked   
...                                   ...           ...         ...   
5987                          Roebuck Bay        Diving  Unprovoked   
5988                              Unknown  Pearl diving  Unprovoked   
5989                       Ocracoke Inlet      Swimming  Unprovoked   
5990                 Panama Bay 8�N, 79�W       Unknown  Unprovoked   
5991  Below the English fort, Trincomalee      Swimming  Unprovoked   

       Case Number  Year                Age  
0     2016.09.18.c  2016            Florida  
1     2016.09.18.b  2016            Florida  
2     2016.09.18.a  2016            Florida  
3       2016.09.17  2016           Victoria  
4       2016.09.15  2016           Victoria  
...            ...   ...                ...  
5987       ND.0005     0  Western Australia  
5988       ND.0004     0  Western Australia  
5989       ND.0003     0     North Carolina  
5990       ND.0002     0            Unknown  
5991       ND.0001     0   Eastern Province  

[5992 rows x 8 columns]